tapchannel/allocation.go

```diff
@@ -0,0 +1,526 @@
+	}
+
+	outputsEqual := bytes.Equal(pkScript, finalPkScript) &&
+		value == int64(a.BtcAmount) && cltv == a.CLTV
+
+	return outputsEqual, nil
+}
+
+// FilterByType returns a filter function that can be used to filter a list of
+// allocations by the given allocation type.
+func FilterByType(allocType AllocationType) func(a *Allocation) bool {
+	return func(a *Allocation) bool {
+		return a.Type == allocType
+	}
+}
+
+// piece is a struct that tracks the currently available and allocated assets
+// for a specific asset ID. It also contains the virtual packet that is being
+// built for each asset ID.
+type piece struct {
+	// assetID is the ID of the asset that is being distributed.
+	assetID asset.ID
+
+	// totalAvailable is the sum of all asset outputs that are available for
+	// distribution per asset ID.
+	totalAvailable uint64
+
+	// allocated is the amount of assets that have been allocated so far.
+	allocated uint64
+
+	// proofs is the list of proofs for the assets that make up a piece.
+	proofs []*proof.Proof
+
+	// packet is the virtual packet that is being built for the asset ID.
+	packet *tappsbt.VPacket
+}
+
+// available returns the amount of assets that are still available for
+// distribution.
+func (p *piece) available() uint64 {
+	return p.totalAvailable - p.allocated
+}
+
+// sortPieces sorts the given pieces by asset ID and the contained proofs by
+// amount and then script key.
+func sortPiecesWithProofs(pieces []*piece) {
+	// Sort pieces by asset ID.
+	sort.Slice(pieces, func(i, j int) bool {
+		return bytes.Compare(
+			pieces[i].assetID[:], pieces[j].assetID[:],
+		) < 0
+	})
+
+	// Now sort all the proofs within each piece by amount and then script
+	// key. This will give us a stable order for all asset UTXOs.
+	for idx := range pieces {
+		sort.Slice(pieces[idx].proofs, func(i, j int) bool {
+			assetI := pieces[idx].proofs[i].Asset
+			assetJ := pieces[idx].proofs[j].Asset
+
+			// If amounts are equal, sort by script key.
+			if assetI.Amount == assetJ.Amount {
+				keyI := assetI.ScriptKey.PubKey
+				keyJ := assetJ.ScriptKey.PubKey
+				return bytes.Compare(
+					keyI.SerializeCompressed(),
+					keyJ.SerializeCompressed(),
+				) < 0
+			}
+
+			// Otherwise, sort by amount, but in reverse order so
+			// that the largest amounts are first.
+			return assetI.Amount > assetJ.Amount
+		})
+	}
+}
+
+// DistributeCoins allocates a set of inputs (extracted from the given input
+// proofs) to virtual outputs as specified by the allocations given. It returns
+// a list of virtual packets (one for each distinct asset ID) with virtual
+// outputs that sum up to the amounts specified in the allocations. The main
+// purpose of this function is to deterministically re-distribute heterogeneous
+// asset outputs (asset UTXOs of different sizes from different tranches/asset
+// IDs) according to the distribution rules provided as "allocations".
+func DistributeCoins(inputs []*proof.Proof, allocations []*Allocation,
```

Comment:
this can be used outside of tap channel context too, correct?

Reply:
Sure. But the main reason to use this is to have a deterministic way of distribution that both parties can use (in LN only the TXID of the transactions are exchanged between the peers, so they need to come up with exactly the same TX down to the last bit). So I think this mostly makes sense in the context of Lightning channels. But it might turn out to be useful for other stuff too.